src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -33,47 +28,56 @@ internal static class WinHttpResponseParser
             var response = new HttpResponseMessage();
             bool stripEncodingHeaders = false;
 
+            // Create a single buffer to use for all subsequent WinHttpQueryHeaders string interop calls.
+            // This buffer is the length needed for WINHTTP_QUERY_RAW_HEADERS_CRLF, which includes the status line
+            // and all headers separated by CRLF, so it should be large enough for any individual status line or header queries.
+            int bufferLength = GetResponseHeaderCharBufferLength(requestHandle, Interop.WinHttp.WINHTTP_QUERY_RAW_HEADERS_CRLF);
+            char[] buffer = new char[bufferLength];
```

Comment:
@justinvp @davidsh @stephentoub I believe this is now allocating twice (or even 4x) the required buffer, for _each_ request:
  * 2x from the following code path: `GetResponseHeaderCharBufferLength` is calling into `WinHttpQueryHeaders` which returns the buffer size in bytes not C# `char` which is 16bit wide.
  * another 2x from double buffering instead of lazily extracting headers as they are requested by the application.

The last point is debatable and would need to be profiled (which, if I have time, I will do part of my perf. work). Lacking the necessary telemetry it's hard to understand if most apps really care about the headers or not.